proxy/src/cache/project_info.rs

```diff
@@ -0,0 +1,435 @@
+        secret_map.insert(user, secret.into());
+        Self {
+            secret: secret_map,
+            allowed_ips: None,
+        }
+    }
+    pub fn new_with_allowed_ips(allowed_ips: Arc<Vec<SmolStr>>) -> Self {
+        Self {
+            secret: std::collections::HashMap::new(),
+            allowed_ips: Some(allowed_ips.into()),
+        }
+    }
+    pub fn get_role_secret(&self, user: &SmolStr, ttl: Option<Duration>) -> Option<AuthSecret> {
+        if let Some(secret) = self.secret.get(user) {
+            if let Some(ttl) = ttl {
+                if secret.created_at.elapsed() > ttl {
+                    return None;
+                }
+            }
+            return Some(secret.value.clone());
+        }
+        None
+    }
+
+    pub fn get_allowed_ips(&self, ttl: Option<Duration>) -> Option<Arc<Vec<SmolStr>>> {
+        if let Some(allowed_ips) = &self.allowed_ips {
+            if let Some(ttl) = ttl {
+                if allowed_ips.created_at.elapsed() > ttl {
+                    return None;
+                }
+            }
+            return Some(allowed_ips.value.clone());
+        }
+        None
+    }
+    pub fn invalidate_allowed_ips(&mut self) {
+        self.allowed_ips = None;
+    }
+    pub fn invalidate_role_secret(&mut self, user: &SmolStr) {
+        self.secret.remove(user);
+    }
+}
+
+/// Cache for project info.
+/// This is used to cache auth data for endpoints.
+/// Invalidation is done by console notifications or by TTL (if console notifications are disabled).
+///
+/// We also store endpoint-to-project mapping in the cache, to be able to access per-endpoint data.
+/// One may ask, why the data is stored per project, when on the user request there is only data about the endpoint available?
+/// On the cplane side updates are done per project (or per branch), so it's easier to invalidate the whole project cache.
+pub struct ProjectInfoCacheImpl {
+    cache: DashMap<SmolStr, EndpointInfo>,
+
+    project2ep: DashMap<SmolStr, HashSet<SmolStr>>,
+    config: ProjectInfoCacheOptions,
+    ttl_enabled: AtomicBool,
+}
+
+impl ProjectInfoCache for ProjectInfoCacheImpl {
+    fn invalidate_allowed_ips_for_project(&self, project_id: &SmolStr) {
+        if let Some(endpoints) = self.project2ep.get(project_id) {
+            for endpoint_id in endpoints.value() {
+                if let Some(mut endpoint_info) = self.cache.get_mut(endpoint_id) {
+                    endpoint_info.invalidate_allowed_ips();
+                }
+            }
+        }
+    }
+    fn invalidate_role_secret_for_project(&self, project_id: &SmolStr, user: &SmolStr) {
+        if let Some(endpoints) = self.project2ep.get(project_id) {
+            for endpoint_id in endpoints.value() {
+                if let Some(mut endpoint_info) = self.cache.get_mut(endpoint_id) {
+                    endpoint_info.invalidate_role_secret(user);
+                }
+            }
+        }
```

Comment:
Need to be careful with the dashmaps here as they are RwLocks. Holding two locks at the same time is a path to deadlocks. Since invalidation is rare I would be fine with this being a more expensive operation and cloning:

```suggestion
        let endpoints = self.project2ep
            .get(project_id)
            .map(|kv| kv.value().clone())
            .unwrap_or_default();
        for endpoint_id in endpoints {
            if let Some(mut endpoint_info) = self.cache.get_mut(endpoint_id) {
                endpoint_info.invalidate_role_secret(user);
            }
        }
```

Same with the above use

Reply:
Thanks, updated.

There is also situation with two locks in `gc` method. 
However, I don't think that would lead to the deadlock, because it's always in order `project2ep` and than `cache`.